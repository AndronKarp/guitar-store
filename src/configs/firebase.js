import firebase from "firebase";

const config = {
  apiKey: "AIzaSyDrF2CjklO0SlyV6Kv7pg7p67a4sLV3qbw",
  authDomain: "project-a0c8e.firebaseapp.com",
  databaseURL: "https://project-a0c8e.firebaseio.com",
  projectId: "project-a0c8e",
  storageBucket: "project-a0c8e.appspot.com",
  messagingSenderId: "686927330962",
  appId: "1:686927330962:web:6892625786538e6588304b",
  measurementId: "G-P6Q64LKV8Q"
};

firebase.initializeApp(config);

const database = firebase.database();
const guitarsRef = database.ref("guitars");
const auth = firebase.auth();
const usersRef = database.ref("users");
auth.getCurrentUser = () => {
  return new Promise((resolve, reject) => {
    const unsubscribe = firebase.auth().onAuthStateChanged(user => {
      unsubscribe();
      resolve(user);
    }, reject);
  });
};

export { guitarsRef, auth, usersRef };
