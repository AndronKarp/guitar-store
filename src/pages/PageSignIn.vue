<template>
  <div class="d-flex flex-fill justify-content-center align-items-center">
    <b-form
      class="d-flex flex-column w-25"
      style="min-width: 280px"
      @submit.prevent="signIn"
    >
      <b-form-group>
        <b-form-input
          placeholder="Enter your e-mail..."
          type="email"
          v-model="$v.form.email.$model"
          :state="authState"
        ></b-form-input>
      </b-form-group>

      <b-form-group>
        <b-form-input
          placeholder="Enter your password..."
          type="password"
          v-model="$v.form.password.$model"
          :state="authState"
        ></b-form-input>
        <b-form-invalid-feedback :state="authState">
          Invalid e-mail or password
        </b-form-invalid-feedback>
      </b-form-group>

      <b-button
        :class="{ 'bg-info': !$v.$invalid }"
        :disabled="$v.$invalid"
        type="submit"
      >
        <b-spinner
          v-if="isFormSubmitting"
          label="Loading..."
          variant="light"
          small
        ></b-spinner
        ><template v-else>Sign In</template>
      </b-button>
    </b-form>
  </div>
</template>

<script>
import { required } from "vuelidate/lib/validators";
import { validationMixin } from "vuelidate";
import { auth } from "../configs/firebase";
import moveToPage from "../mixins/move-to-page";

export default {
  data() {
    return {
      form: {
        email: "",
        password: ""
      },
      authState: null,
      isFormSubmitting: false
    };
  },
  validations: {
    form: {
      email: { required },
      password: { required }
    }
  },
  methods: {
    signIn() {
      this.authState = null;
      this.isFormSubmitting = true;
      auth
        .signInWithEmailAndPassword(this.form.email, this.form.password)
        .then(() => this.moveTo("PageGuitarStore"))
        .catch(() => (this.authState = false))
        .finally(() => (this.isFormSubmitting = false));
    }
  },
  mixins: [validationMixin, moveToPage]
};
</script>

<style></style>
