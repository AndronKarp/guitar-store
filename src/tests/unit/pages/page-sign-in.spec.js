import { mount, createLocalVue } from "@vue/test-utils";
import PageSignIn from "@/pages/PageSignIn";
import BootstrapVue from "bootstrap-vue";
import { auth } from "@/configs/firebase";

const localVue = createLocalVue();

localVue.use(BootstrapVue);

describe("PageSignIn.vue", () => {
  let $router;
  let wrapper;

  beforeEach(() => {
    $router = { push: jest.fn() };
    wrapper = mount(PageSignIn, { localVue, mocks: { $router } });
  });

  test("renders sign in form", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("form submit button", () => {
    let submitButton;
    let inputs;

    beforeEach(() => {
      submitButton = wrapper.find("button");
      inputs = wrapper.findAll("input");
    });

    test("disabled if input data is not provided", () => {
      expect(submitButton.element.getAttribute("disabled")).toBe("disabled");
    });

    test("enabled if input data is provided", () => {
      inputs.setValue("value");
      wrapper.vm.$nextTick(() => {
        expect(submitButton.element.getAttribute("disabled")).toBe(null);
      });
    });
  });

  describe("methods", () => {
    describe("signIn", () => {
      beforeAll(() => {
        auth.signInWithEmailAndPassword = jest
          .fn()
          .mockResolvedValueOnce()
          .mockResolvedValueOnce()
          .mockRejectedValueOnce();
      });

      test("invokes auth.signInWithEmailAndPassword function", async () => {
        await wrapper.vm.signIn();
        expect(auth.signInWithEmailAndPassword).toHaveBeenCalledWith(
          wrapper.vm.form.email,
          wrapper.vm.form.password
        );
      });

      test("invokes moveTo method after successful authorization", async () => {
        const moveTo = jest.spyOn(wrapper.vm, "moveTo");
        await wrapper.vm.signIn();
        expect(moveTo).toHaveBeenCalledWith("PageGuitarStore");
      });

      test("sets authState value to false if authorization is failed", async () => {
        await wrapper.vm.signIn();
        wrapper.vm.$nextTick(() => {
          expect(wrapper.vm.authState).toBe(false);
        });
      });
    });
  });
});
