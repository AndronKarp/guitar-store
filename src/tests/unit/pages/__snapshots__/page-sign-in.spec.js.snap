// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageSignIn.vue renders sign in form 1`] = `
<div class="d-flex flex-fill justify-content-center align-items-center">
  <form novalidate="novalidate" class="d-flex flex-column w-25" style="min-width: 280px;">
    <fieldset class="form-group" id="__BVID__10">
      <!---->
      <div tabindex="-1" role="group" class="bv-no-focus-ring"><input type="email" placeholder="Enter your e-mail..." class="form-control" novalidate="novalidate" id="__BVID__11">
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset>
    <fieldset class="form-group" id="__BVID__12">
      <!---->
      <div tabindex="-1" role="group" class="bv-no-focus-ring"><input type="password" placeholder="Enter your password..." class="form-control" id="__BVID__13">
        <div class="invalid-feedback">
          Invalid e-mail or password
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset> <button type="submit" disabled="disabled" class="btn btn-secondary disabled">Sign In</button>
  </form>
</div>
`;
